### ISSUE TITLE
memory management

### ISSUE THREAD

**Issue by barnabasmolnar** [xeokit/xeokit-sdk#347]
Dear Lindsay,

We have a couple of observations and questions regarding the **memory management** of xeokit.

We've done some digging and it seems that regardless of model size, xeokit always reserves at least ~300MB of memory.

What seems to happen is that a number of arrays are initialised to a huge size as can be seen in https://github.com/xeokit/xeokit-sdk/blob/master/src/viewer/scene/PerformanceModel/lib/batching/BatchingBuffer.js#L5.

Say you have a model that would only be a five kilobytes in size and would contain 4 walls at most; reserving this much memory seems a tad wasteful for that.

What's more, it seems like that when the viewer is destroyed, this memory is **never actually released**. From what we can tell, there's no mechanism here that would set the `freeBuffers` array back to empty when the viewer is being destroyed, so all the stuff already there will not be garbage collected and thus just linger in memory. We did some quick tests, wrote a function to clear the `freeBuffers` array and just called it manually when destroying the viewer. Which seemed to do the trick and reduced the memory footprint by a hefty amount. But of course, it would be best if this was part of the destroy pipeline rather than having to take care of this manually.

This is especially problematic in a single page application when the page is rarely actually reloaded, and can cause issues, slow downs even when the user is no longer on the dedicated model viewer page. We do take care to destroy the viewer when it's no longer rendered but as explained above, it does not free up the memory taken by it.

The other issue/concern we have kind of relates to all of the above. We have a model loading process that is very much **parallel** in nature, that is, we don't wait for one model to load before the other can be loaded and so on. Which means, if we have 10 models, however small they may be, each of them will reserve their own ~300MB of memory adding up to about ~3GB altogether. Not exactly desirable. Would you have some advice/pointers for us on this? We saw in a couple of your examples that you load multiple models in a sequential manner (ie in the `on("loaded")` callbacks) instead. Is that what you would recommend for us as well or is there perhaps room for improvement in terms of support for a parallel model loading approach?

Finally, going back to the first point a bit, this seems a bit too rigid/random as per your comment as wellˇˇ

```js
const MAX_VERTS = SLICING ? (bigIndicesSupported ? 5000000 : 65530) : 5000000;
const MAX_INDICES = MAX_VERTS * 3; // Rough rule-of-thumb
```

Would there perhaps be a way of dynamically allocating resources for individual models? AFAIK, this information should be available in `GLTF` files for instance. @eriadam, would you please share what you showed us today regarding this?

We'd be very keen to have your thoughts on all of the above. Thank you!

**Comment by xeolabs** [comment]
Hi @barnabasmolnar, 

I've published a fix for this in ````xeokit/xeokit-sdk 1.3.42```` - see details in PR https://github.com/xeokit/xeokit-sdk/pull/348https://github.com/xeokit/xeokit-sdk/pull/348

**Comment by barnabasmolnar** [comment]
Sounds cool, thanks! Will have a look as soon as we can. 👍
